clcache.py

```diff
@@ -74,6 +80,80 @@ def __str__(self):
         return repr(self.message)
 
 
+class ManifestsManager(object):
+    HASH_INVALIDATION_COUNTER = 3
+
+    def __init__(self, manifestsRootDir):
+        self._manifestsRootDir = manifestsRootDir
+
+    def manifestDir(self, manifestHash):
+        return os.path.join(self._manifestsRootDir, manifestHash[:2])
+
+    def manifestPath(self, manifestHash):
+        return os.path.join(self.manifestDir(manifestHash), manifestHash + ".json")
+
+    def setManifest(self, manifestHash, manifest):
+        ensureDirectoryExists(self.manifestDir(manifestHash))
+        with open(self.manifestPath(manifestHash), 'w') as outFile:
+            # Convert namedtupel to dict to get key names
+            doc = {
+                'includeFiles': manifest.includeFiles,
+                'includesContentToObjectMap': manifest.includesContentToObjectMap
+            }
+            json.dump(doc, outFile, sort_keys=True, indent=2)
+
+    def getManifest(self, manifestHash):
+        fileName = self.manifestPath(manifestHash)
+        if not os.path.exists(fileName):
+            return None
+        try:
+            with open(fileName, 'r') as inFile:
+                doc = json.load(inFile)
+                return Manifest(doc['includeFiles'], doc['includesContentToObjectMap'])
+        except IOError:
+            return None
+
+    def clean(self, maxManifestsSize):
+        manifestFiles = []
+        for path, _, filenames in WALK(self._manifestsRootDir):
+            for filename in filenames:
+                manifestFiles.append(os.path.join(path, filename))
+
+        manifestFileInfos = []
+        for filepath in manifestFiles:
+            try:
+                manifestFileInfos.append((os.stat(filepath), filepath))
+            except OSError:
+                pass
+
+        manifestFileInfos.sort(key=lambda t: t[0].st_atime, reverse=True)
+
+        currentSize = 0
+        for stat, filepath in manifestFileInfos:
+            currentSize += stat.st_size
+            if currentSize < maxManifestsSize:
+                # skip as long as maximal size not reached
+                continue
+            os.remove(filepath)
+
+    @staticmethod
+    def getManifestHash(compilerBinary, commandLine, sourceFile):
+        compilerHash = getCompilerHash(compilerBinary)
+
+        # NOTE: We intentionally do not normalize command line to include
+        # preprocessor options. In direct mode we do not perform
+        # preprocessing before cache lookup, so all parameters are important
+        additionalData = "{}|{}|{}".format(
+            compilerHash, commandLine, ManifestsManager.HASH_INVALIDATION_COUNTER)
+        return getFileHash(sourceFile, additionalData)
+
+    @staticmethod
+    def getIncludesContentHash(listOfHeaderHashes):
+        hasher = HashAlgorithm()
+        hasher.update(','.join(listOfHeaderHashes).encode("UTF-8"))
+        return hasher.hexdigest()
```

Comment:
Does this help? The line becomes a bit long (87 columns)

```
        return HashAlgorithm(','.join(listOfHeaderHashes).encode('UTF-8')).hexdigest()
```

I do not have a strong opinion for any version here